### PR TITLE
Add a new API getProperty for Augment and RadioCtrlCard

### PR DESCRIPTION
**Pull request by Simba98**
- Add a new API getProperty(string) -- get a property of the stock with name of the property
- This PR dependent on The ImmersiveRailroading PR #658, the IR/entity/EntityRollingStock.java : getName()

### CODE REVIEW

File: src/main/java/cam72cam/immersiverailroading/thirdparty/CommonAPI.java [hunk 1]
```diff
@@ -50,26 +102,122 @@ public FluidStack getFluid() {
...
+                    if (stock instanceof EntityMoveableRollingStock) {
+                        return new Object[] { (EntityMoveableRollingStock)stock.getCurrentSpeed().metric() };
+                    } 
+                    return null;
+                case DIRECTION:
+                    if (stock instanceof EntityMoveableRollingStock) {
+                        double[] SpeedVector = getSpeedVector();
+                        return new Object[] { EnumFacing.fromAngle(SpeedVector[1]).toString()) };
+                    } else {
+                        EnumFacing dir = EnumFacing.fromAngle(stock.getRotationYaw());
```
Comment (cam72cam) [inline src/main/java/cam72cam/immersiverailroading/thirdparty/CommonAPI.java]: I don't think you need the if/else here.  The else condition seems to be correct as the speedVector is generated from the rotation yaw value

File: src/main/java/cam72cam/immersiverailroading/thirdparty/CommonAPI.java [hunk 2]
```diff
@@ -50,26 +102,122 @@ public FluidStack getFluid() {
...
+        return getProperty(PropMap.get(PropertyName));
+    }
+
+    public Object[] getProperty(stockProperty PropertyName) {
+        if (stock != null) {
+            switch(PropertyName) {
+                case ID:
+                    return new Object[] { stock.getDefinitionID() };
+                case NAME:
+                    return new Object[] { stock.getName() };
```
Comment (cam72cam) [inline src/main/java/cam72cam/immersiverailroading/thirdparty/CommonAPI.java]: stock.getDefinition().getName()

File: src/main/java/cam72cam/immersiverailroading/thirdparty/CommonAPI.java [hunk 3]
```diff
@@ -19,6 +19,33 @@
...
 public class CommonAPI {
     private final EntityRollingStock stock;
+    public static final HashMap<String, stockProperty> PropMap;
+    static  
+    {  
+        PropMap = new HashMap<String, stockProperty>();  
+        PropMap.put("id", stockProperty.ID);  
+        PropMap.put("name", stockProperty.NAME);  
+        PropMap.put("tag", stockProperty.TAG);  
+        PropMap.put("weight", stockProperty.WEIGET);
```
Comment (cam72cam) [inline src/main/java/cam72cam/immersiverailroading/thirdparty/CommonAPI.java]: WEIGET -> WEIGHT

Comment (Simba98) [reply src/main/java/cam72cam/immersiverailroading/thirdparty/CommonAPI.java]: :( ... my fault

File: src/main/java/cam72cam/immersiverailroading/thirdparty/CommonAPI.java [hunk 4]
```diff
@@ -39,6 +66,31 @@ public CommonAPI(EntityRollingStock stock) {
         this.stock = stock;
     }
 
+    public Enum stockProperty {
+        ID,
+        NAME,
+        TAG,
+        WEIGET,
```
Comment (Simba98) [inline src/main/java/cam72cam/immersiverailroading/thirdparty/CommonAPI.java]: weight

File: src/main/java/cam72cam/immersiverailroading/thirdparty/CommonAPI.java [hunk 5]
```diff
@@ -39,6 +66,31 @@ public CommonAPI(EntityRollingStock stock) {
         this.stock = stock;
     }
 
+    public Enum stockProperty {
```
Comment (cam72cam) [inline src/main/java/cam72cam/immersiverailroading/thirdparty/CommonAPI.java]: change to StockProperty to match casing in the rest of the project

File: src/main/java/cam72cam/immersiverailroading/thirdparty/CommonAPI.java [hunk 6]
```diff
@@ -50,26 +102,122 @@ public FluidStack getFluid() {
...
+                    return new Object[] { getUniqueID() };   
+                case FLUID:
+                    FluidStack fluid = getFluid();
+                    if (fluid != null) {
+                        return new Object[] { fluid.getFluid().getName(), fluid.amount };
+                    } else {
+                        return null;
+                    }
+                case FLUID_MAX:
+                    return new Object[] { ((FreightTank) stock).getTankCapacity().MilliBuckets() };
```
Comment (Simba98) [inline src/main/java/cam72cam/immersiverailroading/thirdparty/CommonAPI.java]: Here is unsafe , should add stock instanceof FreightTank checking.

File: src/main/java/cam72cam/immersiverailroading/thirdparty/CommonAPI.java [hunk 7]
```diff
@@ -19,6 +19,33 @@
 
 public class CommonAPI {
     private final EntityRollingStock stock;
+    public static final HashMap<String, stockProperty> PropMap;
+    static  
+    {  
+        PropMap = new HashMap<String, stockProperty>();
```
Comment (cam72cam) [inline src/main/java/cam72cam/immersiverailroading/thirdparty/CommonAPI.java]: This map is not needed.  You can interface with the enum like:
name of property:  stockProperty.toString().toLowerCase()
property from name: stockProperty.valueOf(inputName.toUpperCase())

Comment (cam72cam) [reply src/main/java/cam72cam/immersiverailroading/thirdparty/CommonAPI.java]: Could even add helper functions inside the enum definition itself

File: src/main/java/cam72cam/immersiverailroading/thirdparty/CommonAPI.java [hunk 8]
```diff
@@ -50,26 +102,122 @@ public FluidStack getFluid() {
...
+                case DIRECTION:
+                    if (stock instanceof EntityMoveableRollingStock) {
+                        double[] SpeedVector = getSpeedVector();
+                        return new Object[] { EnumFacing.fromAngle(SpeedVector[1]).toString()) };
+                    } else {
+                        EnumFacing dir = EnumFacing.fromAngle(stock.getRotationYaw());
+                        return new Object[] { dir.toString() };
+                    }
+                case SPEED_DIRECTION:
+                    if (stock instanceof EntityMoveableRollingStock) {
```
Comment (cam72cam) [inline src/main/java/cam72cam/immersiverailroading/thirdparty/CommonAPI.java]: Is SPEED_DIRECTION needed?

File: src/main/java/cam72cam/immersiverailroading/thirdparty/CommonAPI.java [hunk 9]
```diff
@@ -50,26 +102,122 @@ public FluidStack getFluid() {
         return null;
     }
 
+    public Object[] getProperty(String PropertyName) {
+        return getProperty(PropMap.get(PropertyName));
+    }
+
+    public Object[] getProperty(stockProperty PropertyName) {
```
Comment (cam72cam) [inline src/main/java/cam72cam/immersiverailroading/thirdparty/CommonAPI.java]: I like your pattern of
if (isinstance) {
  stuff for that instance
}

Let's make sure that's consistent through the function

**Review by cam72cam — CHANGES_REQUESTED** [review]
I've left a bunch of comments on the code and have a few more "broad strokes" ideas / questions once those are fixed.

I definitely think this is going in the right direction and want to see it merged into IR for the next version.

Good work :+1:

File: src/main/java/cam72cam/immersiverailroading/thirdparty/CommonAPI.java (hunk 7 above)
Comment (Simba98) [reply src/main/java/cam72cam/immersiverailroading/thirdparty/CommonAPI.java]: It is used to convert String arguments from Lua script to Enum.
I think maybe the name of properties will be changed in future.
If we have this hashmap, the name and the arguments will be decoupled.
Well, if you insist...

File: src/main/java/cam72cam/immersiverailroading/thirdparty/CommonAPI.java (hunk 8 above)
Comment (Simba98) [reply src/main/java/cam72cam/immersiverailroading/thirdparty/CommonAPI.java]: For compatibility with info calls, if you wish get them either.
I suggest keep it.

**Comment by Simba98** [comment]
> I've left a bunch of comments on the code and have a few more "broad strokes" ideas / questions once those are fixed.
> I definitely think this is going in the right direction and want to see it merged into IR for the next version.
> Good work 👍

Thanks for your reviewing. I'll change them.
And Endal also gave me some suggestions in Discord.
Maybe I can also apply the suggestions.

File: src/main/java/cam72cam/immersiverailroading/thirdparty/CommonAPI.java (hunk 1 above)
Comment (Simba98) [reply src/main/java/cam72cam/immersiverailroading/thirdparty/CommonAPI.java]: I think it is necessary.
In your info() function, when the currentSpeed < 0, your code set the direction opposite through:
 dir = dir.getOpposite();

My function getSpeedVector, do the same thing in the if condition; and fall back to else condition when it don't belong to EntityMoveableRollingStock.

File: src/main/java/cam72cam/immersiverailroading/thirdparty/CommonAPI.java (hunk 7 above)
Comment (cam72cam) [reply src/main/java/cam72cam/immersiverailroading/thirdparty/CommonAPI.java]: If that is a concern, we could change the constructor of the enum to take a string to use as an identifier and do the lookup based on that

File: src/main/java/cam72cam/immersiverailroading/thirdparty/CommonAPI.java (hunk 8 above)
Comment (cam72cam) [reply src/main/java/cam72cam/immersiverailroading/thirdparty/CommonAPI.java]: Ok

File: src/main/java/cam72cam/immersiverailroading/thirdparty/CommonAPI.java [hunk 10]
```diff
@@ -50,26 +83,126 @@ public FluidStack getFluid() {
         return null;
     }
 
+    public Object[] getProperty(String PropertyName) {
+        return getProperty(StockProperty(PropertyName));
```
Comment (cam72cam) [inline src/main/java/cam72cam/immersiverailroading/thirdparty/CommonAPI.java]: This does not compile.

You will need to change the enum to have a getName and fromName method like:
https://onlinegdb.com/SkQDwmI_H

File: src/main/java/cam72cam/immersiverailroading/thirdparty/CommonAPI.java (hunk 1 above)
Comment (cam72cam) [reply src/main/java/cam72cam/immersiverailroading/thirdparty/CommonAPI.java]: Looking at how the code is written, EnumFacing.fromAngle(stock.getRotationYaw()) is the correct value.

Comment (cam72cam) [reply src/main/java/cam72cam/immersiverailroading/thirdparty/CommonAPI.java]: The other code is wrong

**Comment by cam72cam** [comment]
@Simba98 I will be doing a stream later today on how to setup a dev environment where you can test these changes live.

**Comment by Simba98** [comment]
> @Simba98 I will be doing a stream later today on how to setup a dev environment where you can test these changes live.

That's what I need. Thanks.

**Comment by Simba98** [comment]
@cam72cam 
If you suggest that use a for loop for getNames like that:

> public static MyEnum fromName(String name) {
>             for (MyEnum possible : MyEnum.values()) {
>                 if (possible.name.equals(name)) {
>                     return possible;
>                 }
>             }
>             throw new RuntimeException("Invalid Enum Value: " + name);
>  }

the time complexity of algorithm will be O(N), N is the number of properties.
If we use the HashMap, the time complexity will be O(1).

I suggest my origin implementation would be faster than enum.
Although it have a static HashMap so that it cost more memory.

**Comment by cam72cam** [comment]
I don't think the overhead of the for loop will be a problem.  The code is simpler and harder to mess up.

File: src/main/java/cam72cam/immersiverailroading/thirdparty/CommonAPI.java (hunk 1 above)
Comment (Simba98) [reply src/main/java/cam72cam/immersiverailroading/thirdparty/CommonAPI.java]: > The other code is wrong

image
Cam, this is your origin code. 
https://github.com/TeamOpenIndustry/ImmersiveRailroadingIntegration/blob/forge_1.12.2/src/main/java/cam72cam/immersiverailroading/thirdparty/CommonAPI.java#L63
if the metric()<0 , dir = dir.getOpposite();
Maybe it is my misunderstanding, about the getOpposite. What should I do? Delete the if condition and leave the else part alone?

**Comment by Simba98** [comment]
> I don't think the overhead of the for loop will be a problem.  The code is simpler and harder to mess up.

Well, I'm going with the valueOf().

**Pull request closed without merging**